dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxOutputRowsRule.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.sql.watchdog
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.plans.logical.{GlobalLimit, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.Origin
+import org.apache.spark.sql.execution.command.InsertIntoDataSourceDirCommand
+
+import org.apache.kyuubi.sql.KyuubiSQLConf
+
+/*
+* Add MaxOutputRows rule for output rows limitation to avoid huge output rows of query unexpectedly
+* mainly applied to cases as below:
+* case 1:
+*  {{{
+*   INSERT OVERWRITE DIRECTORY (path=STRING)?
+*   USING format OPTIONS ([option1_name "option1_value", option2_name "option2_value", ...])
+*   SELECT ...
+* }}}
+*
+* case 2:
+* {{{
+*   SELECT [c1, c2, ...]
+* }}}
+*
+* case 3:
+* {{{
+*   WITH CTE AS (
+*   ...)
+* SELECT [c1, c2, ...] FROM CTE ...
+* }}}
+*
+* The Logical Rule add a GlobalLimit node before root project
+* */
+case class MaxOutputRowsRule(session: SparkSession) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!plan.resolved) {
+      plan
+    } else {
+      conf.getConf(KyuubiSQLConf.WATCHDOG_MAX_OUTPUTROWS) match {
+        case Some(maxOutputRows) => plan match {
```

Comment:
how about match `plan.maxRows > maxOutputRows` ?

Reply:
It's origin limit the non-limit query, but It cann't cover with existed limit statement, It's easy to misunderstanding the option name，I rename `spark.sql.watchdog.maxOutputRows` to `spark.sql.watchdog.forcedMaxOutputRows`

Reply:
> how about match `plan.maxRows > maxOutputRows` ?

if non-limit statement, plan.maxRows is None type，the rule will match non-limit statement and skip statement with limit

Reply:
we can add limit if:
-  plan.maxRows is none
- plan.maxRows.get > maxOutputRows

Reply:
About this issue

1. process `plan.maxRows is none` that node without limit statement
2. `plan.maxRows.get > maxOutputRows` illustrate exists limit statement in query, so don't cover user choice, or else user query result exceed expectation as a bug, so skip the situation

Reply:
Do you mean if user's query limit 100 and the config is 10 then we can't add the limit ? I think we should add limit as long as `plan.maxRows is none` or `plan.maxRows.get > maxOutputRows`. Otherwise it's hard to say which limit is added by us.

Reply:
> Do you mean if user's query limit 100 and the config is 10 then we can't add the limit ? 
Yes.
> I think we should add limit as long as `plan.maxRows is none` or `plan.maxRows.get > maxOutputRows`. Otherwise it's hard to say which limit is added by us.
Do you mean using maxOutputRows cover user limit setting ？

Reply:
yes, otherwise this config has no meaning

Reply:
> yes, otherwise this config has no meaning

OK, Final limit according to `spark.sql.watchdog.forcedMaxOutputRows`, commit:  https://github.com/apache/incubator-kyuubi/pull/1085/commits/5b0383dd5d46dbb752ecd1bd37fea5f8fbbb60c8